### PR TITLE
Use POST validator statuses in VC

### LINKED ISSUE #7733: use POST validatorStatuses in VC
Now that the POST validatorStatuses is supported in the BN, we should start using it in the VC, but we probably need a feature flag or to fall back to the existing GET if we get a 404 response...

Ref. #7706

### PR DESCRIPTION
**Pull request by StefanBratanov**
## PR Description

- Add postValidators to `ValidatorRestApiClient`
- Change `RemoteValidatorApiHandler` to try **POST** request before doing a batched **GET** request
- Remove unused methods from `OkHttpValidatorRestApiClient` (they have been replaced by typeDef requests)

## Fixed Issue(s)
fixes #7733 

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Review by zilm13 — COMMENTED** [review]
LGTM, but I'm concerned on the spec filters definition
```
The lists of validator IDs and statuses to filter on. Either or both may be **null** to signal that no filtering on that attribute is desired.
```
empty array looks intuitively good for me but here we can get client implementations discrepancy

File: validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java [hunk 1]
```diff
@@ -322,7 +357,7 @@ public Optional<SyncCommitteeContribution> createSyncCommitteeContribution(
             queryParams,
             EMPTY_MAP,
             createHandler(GetSyncCommitteeContributionResponse.class)
-                .withHandler(SC_NOT_FOUND, (request, response) -> Optional.empty()))
+                .withHandler(SC_NOT_FOUND, (req, res) -> Optional.empty()))
```
Comment (rolfyone) [inline validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java]: request response reads better...

File: validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java [hunk 2]
```diff
@@ -127,8 +135,21 @@ public Optional<GetBlockHeaderResponse> getBlockHeader(final String blockId) {
...
+   * Tries <a
+   * href="https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/postStateValidators">POST
+   * Get validators from state</a> and falls back to <a
+   * href="https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getStateValidators">GET
+   * Get validators from state</a> if POST method is not allowed
+   */
   @Override
   public Optional<List<ValidatorResponse>> getValidators(final List<String> validatorIds) {
+    try {
+      return postValidators(validatorIds);
```
Comment (rolfyone) [inline validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java]: this is kind of the wrong place...
Basically getValidators has to split on id, and we don't have to with the post. It'd be better if we're determining to use post and calling for ALL keys in 1 call, rather than this approach, but it could potentially be another PR...
The other side of it is much like the SSZ concept, where if we get a post failure once, we should probably assume the endpoint doesnt support it and stop trying and failing every single time we request, and thats likely another good reason to do this either in a wrapping class, or higher up... the wrapper class could potentially encapsulate the splitting as required too...

Comment (zilm13) [reply validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java]: Great catch, on some test node with 5000 validators we could have 100 bad post calls and 100 good gets when it's not supported and 100 post calls instead of 1 when it's supported.

Comment (StefanBratanov) [reply validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java]: Got it. Working on the changes. :)

**Comment by StefanBratanov** [comment]
> LGTM, but I'm concerned on the spec filters definition
> 
> ```
> The lists of validator IDs and statuses to filter on. Either or both may be **null** to signal that no filtering on that attribute is desired.
> ```
> 
> empty array looks intuitively good for me but here we can get client implementations discrepancy

good point, think setting to null is better

**Comment by StefanBratanov** [comment]
> > LGTM, but I'm concerned on the spec filters definition
> > ```
> > The lists of validator IDs and statuses to filter on. Either or both may be **null** to signal that no filtering on that attribute is desired.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > empty array looks intuitively good for me but here we can get client implementations discrepancy
> 
> good point, think setting to null is better



> > LGTM, but I'm concerned on the spec filters definition
> > ```
> > The lists of validator IDs and statuses to filter on. Either or both may be **null** to signal that no filtering on that attribute is desired.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > empty array looks intuitively good for me but here we can get client implementations discrepancy
> 
> good point, think setting to null is better

Ok, actually just checked, we fail when trying for example request `{"ids":["1"],"statuses":null}` with `Array expected but got VALUE_NULL\n at [Source: REDACTED (StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION disabled); line: 5, column: 14]` so we may be against the spec. Opinions @rolfyone ?

**Comment by StefanBratanov** [comment]
> > > LGTM, but I'm concerned on the spec filters definition
> > > ```
> > > The lists of validator IDs and statuses to filter on. Either or both may be **null** to signal that no filtering on that attribute is desired.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > empty array looks intuitively good for me but here we can get client implementations discrepancy
> > 
> > 
> > good point, think setting to null is better
> 
> > > LGTM, but I'm concerned on the spec filters definition
> > > ```
> > > The lists of validator IDs and statuses to filter on. Either or both may be **null** to signal that no filtering on that attribute is desired.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > empty array looks intuitively good for me but here we can get client implementations discrepancy
> > 
> > 
> > good point, think setting to null is better
> 
> Ok, actually just checked, we fail when trying for example request `{"ids":["1"],"statuses":null}` with `Array expected but got VALUE_NULL\n at [Source: REDACTED (StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION disabled); line: 5, column: 14]` so we may be against the spec. Opinions @rolfyone ? Reverted PR to be an empty list.

**Comment by rolfyone** [comment]
Because Statuses is also an array, if you want to send a 'default' for statuses it isn't null. It's actually an empty array that you would send.

Having said that, because its more like query parameters, the better behaviour is to send just the fields you want to set, and let the server side default them correctly.

if you really wanted to you could do something like `{"ids":[], "statuses":[]}` to request all validator ids, but its much simpler to just send an empty body or `{}`, you'd never send the empty arrays like above...

So TL/DR, we just want to specify `{"ids":[IDS]}` 

We ideally want it without carriage returns - it sounds ugly but it can actually add a lot of size to the request when there's a lot of small numbers.  This would also be worth verifying so we're happy that the request string wont be huge for large requests. The numbers should also be strictly quoted, because many servers are treating them as strings, but teku is far less strict on this front.

File: validator/remote/src/integration-test/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClientTest.java [hunk 3]
```diff
@@ -173,198 +152,46 @@ public void getValidators_WhenSuccess_ReturnsResponse() {
...
-
-    Optional<BeaconBlock> beaconBlock =
-        apiClient.createUnsignedBlock(slot, blsSignature, graffiti, false);
-
-    assertThat(beaconBlock).isPresent();
-    assertThat(beaconBlock.get()).isEqualTo(expectedBeaconBlock);
-  }
-
-  @Test
-  public void createUnsignedBlocks_shouldUseBlindedBlocksIfEnabled() {
```
Comment (rolfyone) [inline validator/remote/src/integration-test/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClientTest.java:239]: I'm not sure why we'd be deleting tests in this PR...

File: validator/remote/src/main/java/tech/pegasys/teku/validator/remote/RemoteValidatorApiHandler.java [hunk 4]
```diff
@@ -127,26 +128,49 @@ public SafeFuture<Map<BLSPublicKey, Integer>> getValidatorIndices(
...
+  private <T> Optional<Map<BLSPublicKey, T>> makeValidatorRequest(
+      final Collection<BLSPublicKey> publicKeys,
+      final Function<ValidatorResponse, T> valueExtractor) {
+    try {
+      return apiClient
+          .postValidators(convertPublicKeysToValidatorIds(publicKeys))
+          .map(responses -> convertToValidatorMap(responses, valueExtractor));
+    } catch (final PostStateValidatorsNotExistingException __) {
+      LOG.debug(
+          "POST method is not available for getting validators from state. Will use GET instead.");
```
Comment (rolfyone) [inline validator/remote/src/main/java/tech/pegasys/teku/validator/remote/RemoteValidatorApiHandler.java]: so here, also remember that there was an exception, and never try again...

Basically each epoch we'd still attempt this post and waste time with the failure if we go this way, where as if we make our first call and that fails, then we move on to gets, the next time we can just do gets... it may just look something like
```
if (isPostWorking) {
    try {
      apiclient.postValidators(....
    } catch () {
      isPostWorking.setAtomicBoolean(false);
      LOG...
    }
}
return makeBatchedValidatorRequest...
```

**Comment by StefanBratanov** [comment]
> Because Statuses is also an array, if you want to send a 'default' for statuses it isn't null. It's actually an empty array that you would send.
> 
> Having said that, because its more like query parameters, the better behaviour is to send just the fields you want to set, and let the server side default them correctly.
> 
> if you really wanted to you could do something like `{"ids":[], "statuses":[]}` to request all validator ids, but its much simpler to just send an empty body or `{}`, you'd never send the empty arrays like above...
> 
> So TL/DR, we just want to specify `{"ids":[IDS]}`
> 
> We ideally want it without carriage returns - it sounds ugly but it can actually add a lot of size to the request when there's a lot of small numbers. This would also be worth verifying so we're happy that the request string wont be huge for large requests. The numbers should also be strictly quoted, because many servers are treating them as strings, but teku is far less strict on this front.

I am not sure how other clients deal with missing fields though. I am afraid that they may fail and would prefer to send [] thus I am trying to be more cautious on this side, because the spec is not precise on that front.

**Update** Reverting on that view. The schema specifies the fields in the json as optional, so it's fine to send just **ids**

File: validator/remote/src/main/java/tech/pegasys/teku/validator/remote/RemoteValidatorApiHandler.java (hunk 4 above)
Comment (StefanBratanov) [reply validator/remote/src/main/java/tech/pegasys/teku/validator/remote/RemoteValidatorApiHandler.java]: Yeah that makes sense. Will implement.

File: validator/remote/src/integration-test/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClientTest.java (hunk 3 above)
Comment (StefanBratanov) [reply validator/remote/src/integration-test/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClientTest.java:239]: Just deleting couple of methods only used in this test in the `ValidatorRestApiClient`

**Review by rolfyone — APPROVED** [review]
Looks good! Thanks for your patience :)
